cEP-0016.md

```diff
@@ -0,0 +1,180 @@
+content.
+
+For example, git revert commit which is used to revert any previous commit and
+git merge commit which is generated whenever someone performs a git rebase and
+tries to merge master onto the pull request. Such commits generate a default
+commit message that is very concise and not according to coala’s standards.
+The coala project wants to allow `git revert` commits, but does not want to
+accept `git merge` commits as the repositories are strictly fast-forward merges
+that do not have an extra commit.
+
+Also in cases where some tests are not required by any pull request made to
+coala, some special sequences can be added to git commit messages to instruct
+tools to operate in a certain way.
+
+This project is about detecting such commits and verifying that they are correct
+according to the configuration settings in .coafile
+
+## Proposed Change
+
+Here is the detailed implementation stepwise:
+
+1. We start by identifying different types of special commits namely revert,
+   merge and commits that disable ci builds.
+2. Each type of commit is then handled as a separate case.
+3. For commits that disable ci builds:
+   - Commit contents are checked to find `[ci skip]`.
+   - For such commits, it is checked that the commit body explains a valid
+     reason why this change should not include ci builds.
+   - Files modified in the commit should include files not covered by `.coafile`
+     or tests. If this is not the case, an error is raised to exclude `[ci skip]`
+     from the commit message.
+4. For merge commits:
+   - Merge commits are identified by checking the parents of the commit.
+   - In case the bear finds that the commit tries to merge master onto the pull
+     request, message is displayed to perform a rebase instead.
+5. For revert commits:
+   - Revert commits are identified based on the format of the commit.
+   - For revert commits, the patches of both commits are compared to ensure that
+     they are in fact reverse of each other. Minor differences are allowed.
+   - In case of total mismatch, a note in the commit message highlighting that
+     its not a clean revert is expected.
+   - A message requesting improvement in the commit message is displayed
+     otherwise.
+
+## Code Prototypes
+
+A prototype for identifying merge commits:
+
+```python
+    def is_merge_commit(self, shortlog, body):
+        commit_hash = run_shell_command('git rev-parse HEAD')
+
+        command = 'git show --no-patch --format="%P" '
+        command += str(commit_hash)
+
+        commit_parents = run_shell_command(command)
+
+        if commit_parents > 1:
+            yield Result(self, 'This is a merge commit.')
+```
+
+A prototype for identifying commits that disable ci builds:
+
+```python
+    def disable_cibuilds(self, shortlog, body):
+        ci_regex = re.compile(r'ci skip')
+
+        if ci_regex.match(shortlog):
+            yield Result(self, 'This commit disables ci builds.')
```

Comment:
why would we yield a `Result` here?

Do you plan on creating a meta bear which analyses commit messages and provides `Result` objects for other bears to use?  That would actually be a very good idea...

It would be very backwards to check the files first and then disable ci-skips..?

We rarely get ci-skip commit messages, and even more rarely would they be acceptable.  Checking the files first is a lot of work, which would be unnecessary 99% of the time.

So as you explain in your Proposed changes above, the bear would first look for ci-skip, and if present it would verify it is permitted based on the files modified.

So this method would not generate an error here, unless this method is also checking the file contents.

As this is only a prototype, you can add a comment in the code to explain where a section of code is not implemented in the prototype, e.g.

```
# Here we would check all of the files modified to ensure that 'ci skip' is permitted
```

Reply:
Most of the tasks for my project are git specific so I don't plan on making a separate meta bear for analyzing commit messages because that part is already done by `CommitBear`. Most of my changes will be in the `GitCommitBear` except for https://github.com/coala/coala-bears/issues/1976 . This is quite generic and so needs a new bear. Altough this won't be a meta bear either but rather a bear dependent on `CommitBear`.